### PR TITLE
issues/163: Some tests in nested classes are not exercised when changing related production code not exercised by the tests' top-level class

### PR DESCRIPTION
**Pull request by pjmartos**
Skippy appears to not be taking into account that test classes may contain nested classes (e.g. JUnit 5 test classes annotated with `@Nested`) that in turn may contain test methods of their own (it is not mandatory, but it is usually the case).

Worts case scenario, if the top-level class itself doesn't exercise a production class that has changed, none of the tests would be executed, even if some of the nested classes of that test class do. The net result is that the change in the production class would be undetected.

### CODE REVIEW

File: skippy-core/src/main/java/io/skippy/core/TestImpactAnalysis.java [hunk 1]
```diff
@@ -134,48 +134,62 @@ PredictionWithReason predict(String testClassName, SkippyConfiguration configura
...
             }
-            var maybeAnalyzedTest = analyzedTests.stream()
-                    .filter(test -> classFileContainer.getById(test.getTestClassId()).getClassName().equals(testClassName))
-                    .findFirst();
-            if (maybeAnalyzedTest.isEmpty()) {
+
+            var testClassIds = getTestClassIds(testClassName);
+            var maybeAnalyzedTests = analyzedTests.stream()
+                    .filter(test -> testClassIds.contains(test.getTestClassId()))
+                    .toList();
```
Comment (pjmartos) [inline skippy-core/src/main/java/io/skippy/core/TestImpactAnalysis.java:141]: Even when working in a mono-module project, we can no longer try to match a single class ID to a single class name, since going forward a class name can have more than one ID (the ID assigned to the top-level class, plus the ID assigned to every nested class contained therein, if any).

**Review by fmck3516 — COMMENTED** [review]
Love this PR. I think that there is one remaining problem that can not be solved with the proposed solution:

```
@PredictWithSkippy
public class LeftPadderTest {

    @Test
    void testSomething() {
        var input = TestConstants.HELLO;
        assertEquals(" hello", LeftPadder.padLeft(input, 6));
    }

    @Nested
    class FirstNestedClass {
        @Test
        void testNothing() {
            assertTrue(true);
        }

    }

}
```

Resetting of the JaCoCo agent in `SkippyTestApi` will result in the loss of the execution data for the free-floating test at the top level once the nested class is executed:
```
    @Test
    void testSomething() {
        var input = TestConstants.HELLO;
        assertEquals(" hello", LeftPadder.padLeft(input, 6));
    }
```

Not sure yet how to solve that.

**Comment by fmck3516** [comment]
I took a stab at dealing with nested test classes: https://github.com/skippy-io/skippy/pull/166/files

Thoughts? I'm also curious to see how you would handle the corner case from my previous comment.

**Comment by pjmartos** [comment]
> Love this PR. I think that there is one remaining problem that can not be solved with the proposed solution:
> 
> ```
> @PredictWithSkippy
> public class LeftPadderTest {
> 
>     @Test
>     void testSomething() {
>         var input = TestConstants.HELLO;
>         assertEquals(" hello", LeftPadder.padLeft(input, 6));
>     }
> 
>     @Nested
>     class FirstNestedClass {
>         @Test
>         void testNothing() {
>             assertTrue(true);
>         }
> 
>     }
> 
> }
> ```
> 
> Resetting of the JaCoCo agent in `SkippyTestApi` will result in the loss of the execution data for the free-floating test at the top level once the nested class is executed:
> 
> ```
>     @Test
>     void testSomething() {
>         var input = TestConstants.HELLO;
>         assertEquals(" hello", LeftPadder.padLeft(input, 6));
>     }
> ```
> 
> Not sure yet how to solve that.

Gah, that's right :D
Thanks for the quick reply, I'll definitely check out your solution!

**Comment by pjmartos** [comment]
> I took a stab at dealing with nested test classes: https://github.com/skippy-io/skippy/pull/166/files
> 
> Thoughts? I'm also curious to see how you would handle the corner case from my previous comment.

Left a couple of minor comments, but overall looks quite elegant, should definitely do the trick :D

**Review by fmck3516 — CHANGES_REQUESTED** [review]
This PR will be obsolete once https://github.com/skippy-io/skippy/pull/166 gets merged into main.
